include/deal.II/hp/refinement.templates.h

```diff
@@ -0,0 +1,208 @@
+                  std::min(min_smoothness_coarsen,
+                           indicators(cell->active_cell_index()));
+              }
+          }
+
+      if (const parallel::TriangulationBase<dim, spacedim> *parallel_tria =
+            dynamic_cast<const parallel::TriangulationBase<dim, spacedim> *>(
+              &dof_handler.get_triangulation()))
+        {
+          max_smoothness_refine =
+            Utilities::MPI::max(max_smoothness_refine,
+                                parallel_tria->get_communicator());
+          min_smoothness_refine =
+            Utilities::MPI::min(min_smoothness_refine,
+                                parallel_tria->get_communicator());
+          max_smoothness_coarsen =
+            Utilities::MPI::max(max_smoothness_coarsen,
+                                parallel_tria->get_communicator());
+          min_smoothness_coarsen =
+            Utilities::MPI::min(min_smoothness_coarsen,
+                                parallel_tria->get_communicator());
+        }
+
+      // Absent any better strategies, we will set the threshold by linear
+      // interpolation for both classes of cells individually.
+      const Number threshold_smoothness_refine =
+                     min_smoothness_refine +
+                     p_refine_fraction *
+                       (max_smoothness_refine - min_smoothness_refine),
+                   threshold_smoothness_coarsen =
+                     min_smoothness_coarsen +
+                     p_coarsen_fraction *
+                       (max_smoothness_coarsen - min_smoothness_coarsen);
+
+      p_adaptivity_from_absolute_threshold(dof_handler,
+                                           indicators,
+                                           threshold_smoothness_refine,
+                                           threshold_smoothness_coarsen,
+                                           compare_refine,
+                                           compare_coarsen);
+    }
+
+
+
+    template <int dim,
+              typename Number,
+              int spacedim,
+              typename ComparatorRefine,
+              typename ComparatorCoarsen>
+    void
+    p_adaptivity_from_criteria(const hp::DoFHandler<dim, spacedim> &dof_handler,
+                               const Vector<Number> &               indicators,
+                               const Vector<Number> &               criteria,
+                               const ComparatorRefine & compare_refine,
+                               const ComparatorCoarsen &compare_coarsen)
+    {
+      AssertDimension(dof_handler.get_triangulation().n_active_cells(),
+                      indicators.size());
+      AssertDimension(dof_handler.get_triangulation().n_active_cells(),
+                      criteria.size());
+
+      std::vector<bool> p_flags(
+        dof_handler.get_triangulation().n_active_cells(), false);
+
+      // TODO: Use std::transform or something equivalent for this task!
```

Comment:
Here and above: What do you have in mind? I think the code is fine as is.

I suspect that one could achieve what you want with `std::reduce` and a custom functor: https://en.cppreference.com/w/cpp/algorithm/reduce But it seems that that function is only available starting with C++17. I think it could also be done with `std::accumulate` with an appropriate functor: https://en.cppreference.com/w/cpp/algorithm/accumulate

Reply:
I thought that there would be an elegant way to set values of an `std::vector` directly by using some iterator based functions from the `algorithm` library to avoid its initialization.

Now that I look at it, the code looks fine to be as well. I think I was too ambitious :-)